File: e2e/java/src/main/java/io/trzcinski/test/user/UserClient.java
package io.trzcinski.test.user;

import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Map;
import org.springframework.lang.Nullable;

import io.trzcinski.test.user.dto.User;

interface UserClient {

    @PostMapping("/user/createWithList")
    void createUsersWithListInput(
        @RequestBody() List<User> payload
    );

    @GetMapping("/user/{username}")
    User getUserByName(
        @RequestParam("username") String username
    );

    @PutMapping("/user/{username}")
    void updateUser(
        @RequestBody() User payload,
        @RequestParam("username") String username
    );

    @DeleteMapping("/user/{username}")
    void deleteUser(
        @RequestParam("username") String username
    );

    @GetMapping("/user/login")
    String loginUser(
        @RequestParam("username") String username,
        @RequestParam("password") String password
    );

    @GetMapping("/user/logout")
    void logoutUser(
    );

    @PostMapping("/user/createWithArray")
    void createUsersWithArrayInput(
        @RequestBody() List<User> payload
    );

    @PostMapping("/user")
    void createUser(
        @RequestBody() User payload
    );
}

File: test-samples/kotlin-retrofit/src/main/kotlin/io/trzcinski/test/user/UserClient.kt
package out.io.trzcinski.test.user

import retrofit2.http.*
import retrofit2.Call

import out.io.trzcinski.test.user.dto.User

interface UserClient {

    @POST("/user/createWithList")
    fun createUsersWithListInput(
    ): Call<Void>

    @GET("/user/{username}")
    fun getUserByName(
        @Path("username") username: String
    ): Call<User>

    @PUT("/user/{username}")
    fun updateUser(
        @Path("username") username: String
    ): Call<Void>

    @DELETE("/user/{username}")
    fun deleteUser(
        @Path("username") username: String
    ): Call<Void>

    @GET("/user/login")
    fun loginUser(
        @Query("username") username: String,
        @Query("password") password: String
    ): Call<String>

    @GET("/user/logout")
    fun logoutUser(
    ): Call<Void>

    @POST("/user/createWithArray")
    fun createUsersWithArrayInput(
    ): Call<Void>

    @POST("/user")
    fun createUser(
    ): Call<Void>
}